### DISCUSSION TITLE
[Possible Bug] FastAPI returning the openapi version as 3.1.0, but it should be 3.0.0

### DISCUSSION THREAD

**Discussion by lagartoxa** [fastapi/fastapi#9789] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
custom_openapi = [{
    "apisSorter": "alpha"
}]

tags_metadata = [ 
    {   
        'name': 'Log',
        'description': 'Software Logs.',
    },  
    {   
        'name': 'Redis',
        'description': 'Volatile DB.',
    },  
]

app = FastAPI(
    title='XXXX DMS API',
    description=(
        'API REST'
        f' | Copyright © {date.today().year}'
        ' All rights reserved'),
    openapi_tags=tags_metadata,
    openapi_schema=custom_openapi,
    docs_url=None,
    swagger_ui_parameters={"defaultModelsExpandDepth": -1},
)
```

### Description

Hi everyone

I may have found a bug. I was using fastapi 0.97.0 for my projects and today when I upgraded it to 0.99.1, I'm getting this error when trying to access the docs page:

<img width="1333" alt="Screenshot 2023-07-03 at 18 18 05" src="https://github.com/tiangolo/fastapi/assets/356524/28f2c1f1-8d9b-4c87-ad74-ae16a7c68642">

As you can see, the image shows that the openapi version returned is 3.1.0 and the error page is telling me that I should use a 3.0.x version.

The image below, shows fastapi 0.97.0 returning the openapi version as 3.0.2, as it should.

<img width="1329" alt="Screenshot 2023-07-03 at 18 17 06" src="https://github.com/tiangolo/fastapi/assets/356524/d334a9b6-bde5-4d88-b215-6b04cd350fba">

Sorry if it isn't a bug, but I couldn't find a proper solution anywhere.

### Operating System

Linux

### Operating System Details

I have a MacBook Pro M1 and I'm running an Ubuntu 20 virtual machine with Parallels on it

### FastAPI Version

0.99.1

### Python Version

3.10.12

**Comment by jgould22** [comment, marked as answer]
I think your swagger-ui.js is cached and the cached one is version 4.

FastAPI uses swagger ui 5 now ( a recent change ) that [supports 3.1](https://github.com/swagger-api/swagger-ui/releases/tag/v5.0.0)

[0.97 uses version 4](https://github.com/tiangolo/fastapi/blob/32935103b12b1548117abef0b4af9dd883898308/fastapi/openapi/docs.py#L20)

Try and make sure that JS is the correct one ([seen here](https://github.com/tiangolo/fastapi/blob/0.99.1/fastapi/openapi/docs.py#L20))

**Reply by lagartoxa** [reply]
Thank you very much, this solved my problem

**Comment by mdegis** [comment]
on the other hand, I need it to be set as `3.0.2` because of some external library. I have set it in the `FastAPI` instance like recommend [here](https://fastapi.tiangolo.com/advanced/extending-openapi/#:~:text=.-,openapi_version,-%3A%20The%20version%20of).

```python
app = FastAPI(
        title="name",
        version=metadata.version("name"),
        docs_url="/api/docs",
        redoc_url="/api/redoc",
        openapi_version="3.0.2",
        openapi_url="/api/openapi.json",
        default_response_class=UJSONResponse,
    )
```

However, when I checked the JSON it is still `3.1.0`

```bash
curl http://localhost:8000/api/openapi.json | jq .openapi
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 57869  100 57869    0     0  5926k      0 --:--:-- --:--:-- --:--:-- 27.5M
"3.1.0"
```

am I missing something?

**Reply by tiangolo** [reply]
Can you please provide a simple example I can copy paste and run here to replicate your problem?

**Reply by mdegis** [reply]
I think openapi version is hardcoded. Maybe we shouldn't be able to change it?


Provided `openapi_version` variable is not used when assinging `self.openapi_version`
https://github.com/tiangolo/fastapi/blob/master/fastapi/applications.py#L116

**Reply by tiangolo** [reply]
Hmm, I'm not sure about it because that would look like FastAPI is gonna generate a different type of output and if it had all the complex logic to generate different versions of OpenAPI at the same time.

Nevertheless, if you know you are not using any of the new things in OpenAPI 3.1.0, and are providing all the required things from 3.0.3, you can hardcode it on your side for now:

```Python
from fastapi import FastAPI

app = FastAPI()
app.openapi_version = "3.0.3"


@app.get("/")
async def read_root():
    return {"message": "Hello World"}
```

Again, this is not fully safe as FastAPI won't change the OpenAPI output, it will output the same OpenAPI 3.1.0 content, you are just hardcoding the version number to workaround other tools that just right away if they don't recognize the version but you know they can handle the parts of OpenAPI that you are outputting.

**Reply by quizmoon** [reply]
Problem is that OpenAPI Client Generators, e.g. openapi-generator have very poor support of 3.1.0 and this fact is forcing to use FastAPI version prior of 0.99.
Possible there are some another client generators with support of 3.1.0, would useful if somebody share information about them and would be great after then update docs here: https://fastapi.tiangolo.com/advanced/generate-clients/

**Reply by mdegis** [reply]
yeap @quizmoon, this is exactly our situation. Frontend  uses [openapi-generator-cli](https://github.com/OpenAPITools/openapi-generator-cli) which has an open issue for [Support for OpenAPI Spec 3.1](https://github.com/OpenAPITools/openapi-generator/issues/9083) since March 2021. 

I wish there was a way to change the version of openapi version fastapi generates but I know that is lots of work.

My current solution is also downgrading the fastapi version to 0.98.0 for now.

**Reply by tiangolo** [reply]
Interesting, I've heard more complaints about client generators not supporting things below 3.1.0 and asking for FastAPI to upgrade for quite a bit, I was waiting mainly for Swagger UI.

For the generators out there that don't support 3.1.0 yet, it's a bit of a chicken-egg problem, generators won't add support for things that are not being used as well.

Nevertheless, most of the changes in OpenAPI 3.1.0 were additions of features or relaxation of constraints. So, in most cases, if you are not using any of the new features, you should be able to do that trick I shared and that would probably work.

Now, @quizmoon are you having problems with the docs at https://fastapi.tiangolo.com/advanced/generate-clients/ ? Is there something not working?

And yes, indeed, if in your research you find more information about the client generators, what doesn't work and, in particular, what does work well, it would be great if you created a question here in the Discussions and answered it yourself with your findings, that would be much appreciated.

**Reply by tiangolo** [reply]
> I wish there was a way to change the version of openapi version fastapi generates but I know that is lots of work.

> My current solution is also downgrading the fastapi version to 0.98.0 for now.

@mdegis wait, didn't the solution I gave you above work? I showed you how to change the version number in the generated OpenAPI. And as it's mostly backwards compatible, that should work in most cases.

Also, you might want to try the other code generator, that is actually the one I recommend in the docs: https://github.com/ferdikoomen/openapi-typescript-codegen , check the docs here: https://fastapi.tiangolo.com/advanced/generate-clients/

**Reply by tiangolo** [reply]
Just for completeness, and just so you know, the main change from OpenAPI 3.0.3 to 3.1.0 is that it is now based on JSON Schema 2020-12, without custom modifications. The previous version had custom modifications.

One of the main differences is the `examples` field, but that wouldn't be used by client generators. The other big difference is that `exclusiveMinimum` and `exclusiveMaximum` was a `bool` in the old spec, but a float in the new ones. Nevertheless, Pydantic has actually always generated JSON Schema in a more recent version where those fields were already `float`s. So, if everything was working, I would think it would keep working now. The fact that it complains about the version check is probably just checking the version and failing early, not that they have known incompatibilities. So, I would think the trick above should work for you. Please try it and let me know afterwards, but try it first.

I don't want to make it a parameter in the `FastAPI` class as that would imply that the generated OpenAPI is different internally, at least for newcomers, and it's just patching the version number. And what I'm showing you is mainly a trick/workaround specifically for you, to deal with the tools you are using. But I don't see a reason why it wouldn't work for you.

Maybe I could even document the trick, and mark it as a "trick". But I would like to hear from you first if it works or not for your apps.

**Reply by quizmoon** [reply]
@tiangolo

So far I could not find any generators which have a announced support of 3.1.0 specs. But if somebody knows one would be great to know also and then update docs. I will make a question in discussions as recommended trying to collect that information.

Just now I have tested the "trick" on 0.99.1 with hardcoded openapi_version="3.0.2" for my app. Openapi-generator generated the same typescript-axios client files. It looks in my case as working workaround.

Thank you.

**Reply by tiangolo** [reply]
Great @quizmoon! Glad to hear it worked for you.

**Reply by mdegis** [reply]
openapi-generator-cli throws some errors first, however adding `--skip-validate-spec` generates the client 👍 thank you @tiangolo, the trick is working.

**Reply by tiangolo** [reply]
Great! 🚀

**Reply by abrahan-munro** [reply]
@tiangolo 
> So, if everything was working, I would think it would keep working now.

Is this still the case after the change to pydantic2?

**Reply by johnthagen** [reply]
Wanted to note that another client generator that doesn't yet support OpenAPI 3.1.0 is [`openapi-python-client`](https://github.com/openapi-generators/openapi-python-client)

- https://github.com/openapi-generators/openapi-python-client/discussions/818

If it were possible to officially output an OpenAPI 3.0.x schema from FastAPI, that'd be great, but can understand why that would be challenging to maintain.

**Reply by masalim2** [reply]
We ran into the same issues with OpenAPI Generator.  Previously correct types in the generated `typescript-axios` client regressed to `any`, and Enums were getting generated as empty interfaces.  

I can confirm this was fixed when using `openapitools/openapi-generator-cli:v7.0.1` with `--skip-validate-spec` and fastapi 0.104 with the `openapi_version` artificially pinned to 3.0.3.

**Comment by JoaquimEsteves** [comment]
This is pretty big annoyance for us at my $WORK

We're using some `AWS` shenanigans to deploy our application and it turns out that their templates don't support the OpenAPI version 3.1.0

We've had to resort to a janky handmade conversion script I cobbled together.

Here's what I have so far:

```python
import typing as t
Json = dict[str | t.Literal["anyOf", "type"], "Json"] | list["Json"] | str | bool


def convert_3_dot_1_to_3_dot_0(json: dict[str, Json]):
    """Will attempt to convert version 3.1.0 of some openAPI json into 3.0.2

    Usage:

        >>> from pprint import pprint
        >>> json = {
        ...     "some_irrelevant_keys": {...},
        ...     "nested_dict": {"nested_key": {"anyOf": [{"type": "string"}, {"type": "null"}]}},
        ...     "examples": [{...}, {...}]
        ... }
        >>> convert_3_dot_1_to_3_dot_0(json)
        >>> pprint(json)
        {'example': {Ellipsis},
         'nested_dict': {'nested_key': {'anyOf': [{'type': 'string'}],
                                        'nullable': True}},
         'openapi': '3.0.2',
         'some_irrelevant_keys': {Ellipsis}}
    """
    json["openapi"] = "3.0.2"

    def inner(yaml_dict: Json):
        if isinstance(yaml_dict, dict):
            if "anyOf" in yaml_dict and isinstance((anyOf := yaml_dict["anyOf"]), list):
                for i, item in enumerate(anyOf):
                    if isinstance(item, dict) and item.get("type") == "null":
                        anyOf.pop(i)
                        yaml_dict["nullable"] = True
            if "examples" in yaml_dict:
                examples = yaml_dict["examples"]
                del yaml_dict["examples"]
                if isinstance(examples, list) and len(examples):
                    yaml_dict["example"] = examples[0]
            for value in yaml_dict.values():
                inner(value)
        elif isinstance(yaml_dict, list):
            for item in yaml_dict:
                inner(item)

    inner(json)
 ```

**Reply by wimvc1** [reply]
Exactly what I was looking for!

Add the following `if` before the first `for` loop to handle `Enum` or `Literal` with only 1 value:
```python     
if "const" in yaml_dict:
   yaml_dict["enum"] = [yaml_dict["const"]]
    del yaml_dict["const"]
```

**Reply by JimmyDore** [reply]
Exactly what I needed, you saved my day, thanks 🙏 
Integrated perfectly into my `custom_openapi` function : 

```python
def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(...)
    convert_3_dot_1_to_3_dot_0(openapi_schema) # 🙏🙏🙏

    # https://github.com/tiangolo/fastapi/issues/4959
    for model in ["FieldSelector"]:
        openapi_schema["components"]["schemas"][model]["oneOf"] = openapi_schema[
            "components"
        ]["schemas"][model].pop("anyOf")

    app.openapi_schema = openapi_schema
    return app.openapi_schema

app.openapi = custom_openapi
```

**Reply by champtar** [reply]
I also needed
```python
if isinstance(yaml_dict.get("exclusiveMinimum"), int):
    yaml_dict["minimum"] = yaml_dict["exclusiveMinimum"]
    yaml_dict["exclusiveMinimum"] = True
if isinstance(yaml_dict.get("exclusiveMaximum"), int):
    yaml_dict["maximum"] = yaml_dict["exclusiveMaximum"]
    yaml_dict["exclusiveMaximum"] = True
```

**Reply by MCFreddie777** [reply]
Hi guys - just stumbled upon this dicussion:

You might wanna check [this official microsoft thing](https://github.com/microsoft/OpenAPI.NET/tree/main/src/Microsoft.OpenApi.Hidi) (but it's a .NET) or this python package: https://github.com/RajeshRoy4426/openapi_downgrade_3_0
